test/metabase/query_processor/middleware/fetch_source_query_test.clj

```diff
@@ -236,3 +236,35 @@
                            (str "Failed to save Card:" e)))]
         (or save-error
             (resolve-card-id-source-tables (circular-source-query card-1-id)))))))
+
+;; Alow complex dependency topologies such as:
+;;
+;;   A
+;:   | \
+;;   B  |
+;;   | /
+;;   C
+;;
+(expect
+  (tt/with-temp* [Card [{card-1-id :id} {:dataset_query {:database (data/id)
+                                                         :type     :query
+                                                         :query    {:source-table (data/id :venues)}}
+                                         :name "foo"
+                                         :display :table
+                                         :visualization_settings {}
+                                         :creator_id 1}]
+                  Card [{card-2-id :id} {:dataset_query {:database (data/id)
+                                                         :type     :query
+                                                         :query    {:source-table (str "card__" card-1-id)}}
+                                         :name "foo"
+                                         :display :table
+                                         :visualization_settings {}
+                                         :creator_id 1}]]
+    (some?
+     (resolve-card-id-source-tables
+      {:database (data/id)
+       :type     :query
+       :query    {:source-table (str "card__" card-1-id)
+                  :joins        [{:alias "c"
+                                  :source-table (str "card__" card-2-id)
+                                  :condition [:= [:field-literal "ID" :type/Number] [:joined-field "c" [:field-id (data/id :venues :id)]]]}]}}))))
```

Comment:
Actually you can just write the whole test like this

```clj
(expect
  (tt/with-temp* [Card [{card-1-id :id} {:dataset_query (data/mbql-query venues)}]
                  Card [{card-2-id :id} {:dataset_query (data/mbql-query nil
                                                          {:source-table (str "card__" card-1-id)})}]]
    (resolve-card-id-source-tables
     (data/mbql-query nil
       {:source-table (str "card__" card-1-id)
        :joins        [{:alias "c"
                        :source-table (str "card__" card-2-id)
                        :condition [:= *ID/Number &c.venues.id]}]}))))
```

I tricked out the `data/mbql-query` and related macros, time will tell whether this was a good idea or not but when working on EE which doesn't have the new version yet I found myself missing it.

`*field/type` = field literal
`&alias.field` = wrap field in `:joined-field`
`!unit.field` = wrap field in `:datetime-field`

Reply:
I also tweaked the macro too to expand almost 100% when you macroexpand-1 it (or `C-c RET` in CIDER) to make it easier to debug/understand